### PR TITLE
Add function and example how to use times instead of frame slices

### LINKED ISSUE #555: FEATURE: Add function/example how to use times instead of frame slices
### Is your feature request related to a problem? Please describe.

To limit the analysis to a couple of frames, we currently allow slicing as an array. However, it would be nice to provide a function that users can use to convert a time into the corresponding frame.

### Describe the solution you'd like

convert something like "1ps" or "1 ps" or "5 ns" into the frame based on the timestep of the trajectory. There is already a helper function in mdacli that might be useful

https://github.com/MDAnalysis/mdacli/blob/99e34787e89a00e243c77cfe3f7826f08f119a38/src/mdacli/utils.py#L47

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by leocle2010**
Fixes #555 

Changes made in this Pull Request:

 - Add function split_time_unit in lib/util.py : Split time and units from string like '12ps'
 - Add function convert_str_time in lib/util.py : Convert a string `time` into a frame number based on given `dt`
 - Add function convert_str_timedict  in lib/util.py : Convert 3 strings 'start', 'stop' and 'step' into there frame number equivalent based on given dt (thought for .run for example)
 - Add an example in examples/basics/advanced-usage.py 

## PR Checklist

 - [X] Docs?
 - [X] Issue raised/referenced?

## LLM / AI generated code disclosure

 - [ ] LLMs or other AI-powered tools (beyond simple code suggestions e.g. inside IDEs) were used in this contribution.

----
📚 Documentation preview 📚: https://maicos--585.org.readthedocs.build/en/585/

### CODE REVIEW

**Comment by hejamu** [comment]
Did we copy the code from mdacli verbatim? Since we have mdacli as a dependency anyway we could just import the methods from there and write a simple wrapper

**Comment by hejamu** [comment]
I changed the code to reuse the mdacli code. We should add a test to check the behaviour.

**Comment by DamienTOQUER** [comment]
We don't use the "convert_str_time" from mdacli as it doesn't work for some edge cases. Typically convert_str_time("1ps",0.1) gives 9 when 10 is expected. It is because in the background (1//0.1) is calculated, and give 9 because of floating point errors. Instead we compute round(1/0.1).

**Comment by DamienTOQUER** [comment]
> We don't use the "convert_str_time" from mdacli as it doesn't work for some edge cases. Typically convert_str_time("1ps",0.1) gives 9 when 10 is expected. It is because in the background (1//0.1) is calculated, and give 9 because of floating point errors. Instead we compute round(1/0.1).

Pull request in mdacli https://github.com/MDAnalysis/mdacli/pull/138

File: examples/basics/advanced-usage.py [hunk 1]
```diff
@@ -22,6 +22,7 @@
 import MDAnalysis as mda
 
 import maicos
+import maicos.lib.util as util
```
Comment (hejamu) [inline examples/basics/advanced-usage.py]: Move this down to where the method is actually used so people know what they have to import.

I would also opt to do a `from maicos.lib.util import convert_str_framedict` for readability.

File: examples/basics/advanced-usage.py [hunk 2]
```diff
@@ -219,3 +220,17 @@
 # 4, 6, 8, and 10, use the ``start``, ``stop``, and ``step`` keywords as follow:
 
 dplan = maicos.DensityPlanar(group_H2O).run(start=10, stop=20, step=2)
+
+
+# %%
+# If you prefer to use times instead of frame slice, the function convert_str_framedict
```
Comment (hejamu) [inline examples/basics/advanced-usage.py]: Looking at it now, the name is not really meaningful to the user. Since this is only a wrapper, we could name it something readable like `times_to_frames()` or `resolve_frames( )`

File: src/maicos/lib/util.py [hunk 3]
```diff
@@ -10,6 +10,7 @@
 import functools
 import inspect
 import logging
+import math
```
Comment (hejamu) [inline src/maicos/lib/util.py]: Don't

File: src/maicos/lib/util.py [hunk 4]
```diff
@@ -869,3 +871,68 @@ def get_module_input_str(module_obj):
...
+
+    Returns
+    -------
+    int
+        frame number
+
+    """
+    val, unit = split_time_unit(x)
+    val_in_ps = mda.units.convert(val, unit, "ps")
+    return round(val_in_ps / dt)
```
Comment (hejamu) [inline src/maicos/lib/util.py]: This goes away once the mdacli fix is through

File: src/maicos/lib/util.py [hunk 5]
```diff
@@ -869,3 +871,68 @@ def get_module_input_str(module_obj):
...
+
+
+def convert_str_framedict(start: str, stop: str, step: str, dt: float) -> dict:
+    """Convert 'start', 'stop', 'step' into their frame number based on given dt.
+
```
Comment (hejamu) [inline src/maicos/lib/util.py:909]: Add some comment about cases are handle.

File: src/maicos/lib/util.py [hunk 6]
```diff
@@ -869,3 +871,68 @@ def get_module_input_str(module_obj):
...
+    -------
+    dict
+        Dictionnary containing start, stop and step frames
+
+    """
+    val, unit = split_time_unit(step)
+    val_in_ps = mda.units.convert(val, unit, "ps")
+    temp_step = round(val_in_ps / dt)
+
+    if not (math.isclose(temp_step * dt, val_in_ps)):
```
Comment (hejamu) [inline src/maicos/lib/util.py]: use np.isclose

```suggestion
    if not (np.isclose(temp_step * dt, val_in_ps)):
```

File: src/maicos/lib/util.py [hunk 7]
```diff
@@ -869,3 +871,68 @@ def get_module_input_str(module_obj):
...
+        raise ValueError("Timestep should be a multiple of dt.")
```
Comment (hejamu) [inline src/maicos/lib/util.py]:
```suggestion
        raise ValueError("step should be a multiple of dt.")
```
You call it step in the input, the error should reuse the naming convention

File: src/maicos/lib/util.py [hunk 8]
```diff
@@ -869,3 +871,68 @@ def get_module_input_str(module_obj):
...
+
+    temp_start = corrected_str_frame(start, dt)
+    temp_stop = corrected_str_frame(stop, dt)
+
+    if not (temp_start < temp_stop):
+        raise ValueError("'start' should be < to 'stop - step'.")
```
Comment (hejamu) [inline src/maicos/lib/util.py]: I am not sure about this. A numpy indexing would just return an empty array, we might want to return the equivalent of that. That means running an analysis like`ana_obj.run(**convert_str_framedict("20ps","10ps","1ps","1ps")` is just the same as `ana_obj.run(start=20, stop=10)` and does nothing.

File: CHANGELOG.rst [hunk 9]
```diff
@@ -11,8 +11,9 @@ Changelog
 
 Unreleased
 ----------
-Alexander Schlaich, Henrik Stooß, Anirban Dutta, Philip Loche, Kira Fischer
+Alexander Schlaich, Henrik Stooß, Anirban Dutta, Philip Loche, Kira Fischer, Leo Legrand
 
+- Add function and example how to use times instead of frame slices (#555)
```
Comment (hejamu) [inline CHANGELOG.rst]:
```suggestion
- Add function and example how to use times instead of frame slices (#585)
```

**Comment by leocle2010** [comment]
Moving back to convert_str_time from mdacli, tests are not passed. Once the release of mdacli is out, this should be solved

**Review by hejamu — APPROVED** [review]
